Optimize txo lookup

shesek opened the pull request:
From https://github.com/facebook/rocksdb/wiki/MultiGet-Performance:

> There is a lot of complexity in the underlying RocksDB implementation to lookup a key. The complexity results in a lot of computational overhead, mainly due to cache misses when probing bloom filters, virtual function call dispatches, key comparisons and IO. Users that need to lookup many keys in order to process an application level request end up calling Get() in a loop to read the required KVs. 
>
> By providing a MultiGet() API that accepts a batch of keys, it is possible for RocksDB to make the lookup more CPU efficient by reducing the number of virtual function calls and pipelining cache misses. Furthermore, latency can be reduced by doing IO in parallel.

shesek committed:
Optimize lookup_txos using rocksdb's MultiGet

See https://github.com/facebook/rocksdb/wiki/MultiGet-Performance

shesek committed:
Optimize and refactor lookup_txos()

- Avoid unnecessary copying of prev outpoints

- When looking for both mempool and on-chain txos, accumulate the set of
  outpoints that remain to be looked up to avoid re-checking for them
  later again in the found set

- Refactored lookup_txos() to use lookup_txo() internally rather than
  the other way around, which was less efficient

- Lookup txos in mempool first, then on-chain

- ChainQuery::lookup_txos() now returns a Result instead of panicking
  when outpoints are missing

- Removed ChainQuery::lookup_avail_txos() and allow_missing, which are
  no longer neceesary

jamesdorfman commented:
These optimizations make sense to me.
utACK 5913e22ae4be667458c35d3b23cdc3e381bfe5b6.

shesek commented:
Rebased.

shesek committed:
Refactor lookup_txos() to use zip()

Following https://github.com/Blockstream/electrs/pull/89#discussion_r1632706930

RCasatta commented on src/new_index/mempool.rs:
```diff
@@ -397,34 +397,29 @@ impl Mempool {
     }
 
-    pub fn lookup_txos(&self, outpoints: &BTreeSet<OutPoint>) -> Result<HashMap<OutPoint, TxOut>> {
+    pub fn lookup_txos(&self, outpoints: BTreeSet<OutPoint>) -> Result<HashMap<OutPoint, TxOut>> {
```
At this stage would be probably okay to have unsorted `outpoings: Vec<OutPoint>` and sorting it out only before hitting db ?
To eventually consider for a follow-up

RCasatta commented:
utACK 4f037c4d9885215ce73de53003756d2e64ec766e

RCasatta commented:
Needs rebase
